src/Microsoft.AspNetCore.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -62,17 +80,39 @@ public MemoryPoolIterator ProducingStart()
 
         public void ProducingComplete(MemoryPoolIterator end)
         {
-            var block = _producingBlock;
-            while (block != end.Block)
+            lock (_writeLock)
             {
-                _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                ProducingCompleteAsync(end).GetAwaiter().GetResult();
```

Comment:
I wonder if it's better not to wait for the writes to complete in `ProducingComplete`. The real `SocketOutput.ProducingComplete` doesn't block.